cpp/ql/lib/experimental/semmle/code/cpp/semantic/analysis/RangeAnalysisStage.qll

```diff
@@ -1095,4 +1109,100 @@ module RangeStage<DeltaSig D, BoundSig<D> Bounds, LangSig<D> LangParam, UtilSig<
   ) {
     bounded(rem.getRightOperand(), b, delta, upper, fromBackEdge, origdelta, reason)
   }
+
+  /**
+   * Define `cmp(true) = <=` and `cmp(false) >=`.
+   *
+   * Holds if `mul = left * right`, and in order to know if `mul cmp(upper) B + k` (for
+   * some `B` and `k`) we need to know that `left cmp(upperLeft) B1 + k1` and
+   * `right cmp(upperRight) B2 + k2` (for some `B1`, `B2`, `k1`, and `k2`).
+   */
+  pragma[nomagic]
+  private predicate boundedMulOperandCand(
```

Comment:
Ooh! I like this predicate a lot - it makes it the different cases quite clear. The qldoc is perhaps slightly inaccurate, since the mentioned `B`s are all zero.

Reply:
Thanks! Yeah, maybe I should replace all those `B`s with `Z`s.

Reply:
Fixed in fe487a1ae80746b4181fd1d40e7fdae46db4995d.

Reply:
How about just `0`? It's perhaps slightly unclear that `Z` means zero (although I haven't checked if this is terminology that's used elsewhere in this file.

Reply:
Hm, yeah maybe `0` is better. That's the `toString` we use on the bound anyway, so that's probably more fitting. I just took the `Z` from my notes with you 😂. I'll replace it with `0`.

Reply:
Fixed in 09d0385f0f2f44cc97e119190d6a624678113410.